src/main/java/com/mraof/minestuck/item/weapon/ItemRightClickEffect.java

```diff
@@ -75,17 +88,110 @@ else if (!world.isRemote)
+							world.addParticle(particle.get(), true, vecPos.x + randomParticleOffsetX, vecPos.y + randomParticleOffsetY, vecPos.z + randomParticleOffsetZ, 0.0D, 0.0D, 0.0D);
+						}
+					}
+				}
+				
+				if(!world.getBlockState(blockPos).allowsMovement(world, blockPos, PathType.LAND))
+				{
+					i = distance * 2;
+				}
+				
+				AxisAlignedBB axisAlignedBB = new AxisAlignedBB(blockPos);
+				LivingEntity closestTarget = player.world.getClosestEntityWithinAABB(LivingEntity.class, visiblePredicate, player, vecPos.x, vecPos.y, vecPos.z, axisAlignedBB);
+				if(closestTarget != null && player instanceof ServerPlayerEntity)
+				{
+					int playerRung = PlayerSavedData.getData((ServerPlayerEntity) player).getEcheladder().getRung();
+					
+					if(closestTarget instanceof UnderlingEntity)
+						closestTarget.attackEntityFrom(DamageSource.causePlayerDamage(player).setMagicDamage(), damage + playerRung / 5F);
+					else
+						closestTarget.attackEntityFrom(DamageSource.causePlayerDamage(player).setMagicDamage(), damage + playerRung / 10F);
+					if(effect != null)
+						closestTarget.addPotionEffect(effect.get());
+					closestTarget.setRevengeTarget(player);
+				} else if(closestTarget != null)
+				{
+					i = distance * 2;
+				}
+				
+				Vec3d particleVecMod = player.getLookVec().inverse().scale(0.5D);
+				Vec3d particleVecPos = new Vec3d(vecPos.x + particleVecMod.x, vecPos.y + particleVecMod.y, vecPos.z + particleVecMod.z);
+				if(i == distance * 2 && explosiveFinish)
+				{
+					world.playSound(null, blockPos, SoundEvents.ENTITY_SHULKER_BULLET_HIT, SoundCategory.BLOCKS, 1.2F, 0.6F);
+					
+					world.addParticle(ParticleTypes.FLASH, particleVecPos.x, particleVecPos.y, particleVecPos.z, 0.0D, 0.0D, 0.0D);
+					for(int a = 0; a < 25 + player.getRNG().nextInt(10); a++)
+					{
+						world.addParticle(particle.get(), true, particleVecPos.x + randomParticleOffsetX, particleVecPos.y + randomParticleOffsetY, particleVecPos.z + randomParticleOffsetZ, player.getRNG().nextGaussian() * 0.12D, player.getRNG().nextGaussian() * 0.12D, player.getRNG().nextGaussian() * 0.12D);
+					}
+				} else if(i == distance * 2)
+				{
+					for(int a = 0; a < 15 + player.getRNG().nextInt(10); a++)
+					{
+						world.addParticle(ParticleTypes.CRIT, true, particleVecPos.x + randomParticleOffsetX, particleVecPos.y + randomParticleOffsetY, particleVecPos.z + randomParticleOffsetZ, player.getRNG().nextGaussian() * 0.2D, player.getRNG().nextGaussian() * 0.2D, player.getRNG().nextGaussian() * 0.2D);
+					}
+				}
+			}
+			
+			if(player.isCreative())
+				player.getCooldownTracker().setCooldown(itemStackIn.getItem(), 10);
+			else
+				player.getCooldownTracker().setCooldown(itemStackIn.getItem(), 50);
+			
+			player.swing(hand, true);
+			itemStackIn.damageItem(6, player, playerEntity -> playerEntity.sendBreakAnimation(Hand.MAIN_HAND));
+			player.addStat(Stats.ITEM_USED.get(itemStackIn.getItem()));
+			return ActionResult.resultPass(itemStackIn);
+		};
+	}
```

Comment:
Gonna read through this implementation in detail sometime later. But for now, could you look at potential ways of making the code in large more readable? Primarily I'm thinking of breaking out parts of the code into appropriately-named sub-functions, or failing that, comment what each code block is meant to do. (if you do break out some parts of the function, it might make sense to move out the magic effect as its own class file to hide away those new internal functions, and reduce clutter in this file)

Reply:
Ended up doing basically everything you suggested at once! Created a separate class to reduce business, split it into several functions, and commented to explain steps